### PR TITLE
feat(middleware): add `content` option (`options.content.filter`, `options.content.transform`)

### PR DESCRIPTION
**Pull request by ferdinando-ferreira**
**What kind of change does this PR introduce?**
Feature

**Did you add tests for your changes?**
Yes

**Summary**
It's a refactoring of the functionality proposed at #330 with the requested changes implemented

### CODE REVIEW

File: lib/middleware.js [hunk 1]
```diff
@@ -65,6 +65,18 @@ module.exports = function wrapper(context) {
...
+          filename,
+          notFound
+        });
+        if (context.options && context.options.content && context.options.content.filter) {
+          if (!context.options.content.filter(currentRequest)) {
+            return goNext();
+          }
+        }
         if (notFound) {
           return resolve(goNext());
```
Comment (ferdinando-ferreira) [inline lib/middleware.js]: Question: why is it that every other "goNext" is a regular return but this one is a return resolve(goNext())?

File: lib/middleware.js [hunk 2]
```diff
@@ -77,36 +77,51 @@ module.exports = function wrapper(context) {
...
-        // server content
-        let content = context.fs.readFileSync(filename);
-        content = handleRangeHeaders(content, req, res);
-
-        let contentType = mime.getType(filename);
+        const defaultTransform = (localRequest, fs) => new Promise((localResolve) => {
+          if (notFound) {
+            return localResolve(null);
+          }
+          fs.readFile(localRequest.filename, (err, content) => {
```
Comment (ferdinando-ferreira) [inline lib/middleware.js]: Is it a problem that the readFileSync was replaced by an async one?

**Comment by ferdinando-ferreira** [comment]
Missing: changing the README.md, working on it but it would gain some time to request comments on the code itself meanwhile.

File: test/tests/server.js [hunk 3]
```diff
@@ -152,6 +152,114 @@ describe('Server', () => {
...
   });
 
+  describe('filters', () => {
+    before((done) => {
+      app = express();
+      const compiler = webpack(webpackConfig);
+      instance = middleware(compiler, {
+        stats: 'errors-only',
+        content: {
+          filter(currentRequest) {
```
Comment (michael-ciniawsky) [inline test/tests/server.js]: `currentRequest` => `res || request`

File: lib/middleware.js [hunk 4]
```diff
@@ -38,6 +38,7 @@ module.exports = function wrapper(context) {
     return new Promise(((resolve) => {
       handleRequest(context, filename, processRequest, req);
       function processRequest() {
+        let notFound;
```
Comment (michael-ciniawsky) [inline lib/middleware.js]: Please elaborate further on the need for this

File: lib/middleware.js [hunk 5]
```diff
@@ -60,37 +61,67 @@ module.exports = function wrapper(context) {
...
+            res.setHeader('Content-Length', content.length);
 
-        const { headers } = context.options;
-        if (headers) {
-          for (const name in headers) {
-            if ({}.hasOwnProperty.call(headers, name)) {
-              res.setHeader(name, context.options.headers[name]);
+            const { headers } = context.options;
+            if (headers) {
+              for (const name in headers) {
```
Comment (michael-ciniawsky) [inline lib/middleware.js]: `name` => `header`

File: lib/middleware.js [hunk 6]
```diff
@@ -60,37 +61,67 @@ module.exports = function wrapper(context) {
...
+              localResolve(content);
+            }
+          });
+        });
+        const optionTransform = (context.options && context.options.content) ? context.options.content.transform : null;
+        return defaultTransform(currentRequest, context.fs)
+          .then(content => ((!optionTransform) ? content : optionTransform(currentRequest, content, context.fs)))
+          .then((content) => {
+            if (content == null) {
+              return Promise.reject();
```
Comment (michael-ciniawsky) [inline lib/middleware.js]: Should we better reject with an `{Error}` here ?

**Comment by michael-ciniawsky** [comment]
> Missing: changing the README.md, working on it but it would gain some time to request comments on the code itself meanwhile.

Yep, let's discuss the implementation first and once approve you can add the docs then :)

File: lib/middleware.js (hunk 1 above)
Comment (michael-ciniawsky) [reply lib/middleware.js]: Good question but I don't know the answer, try to remove the `resolve` and see what's happening (if something happens). Please add a comment that this is something to investigate and clearify :)

**Comment by evilebottnawi** [comment]
@michael-ciniawsky can we describe this feature in main post briefly, it is hard to read a lot of posts?

**Comment by ferdinando-ferreira** [comment]
> can we describe this feature in main post briefly, it is hard to read a lot of posts?

Here is the shortest summary:
 * `options.content.filter`: receives a function with the following arguments: `(request)`. This function is called once for each resource being served by webpack-dev-middleware. Returning `false` makes `webpack-dev-middleware` not to serve said file and pass it along to the next middleware.
 * `options.content.transform`: receives a function with the following arguments: `(request, content, memoryFs)`. This function is called once for each resource being served by webpack-dev-middleware and must return a promise whose `resolve` contains the Buffer with the actual content to be served and whose `reject` makes `webpack-dev-middleware` not to serve said file and pass it along to the next middleware.

The added tests contain examples of use cases for these options and, in general, they supercede what's being offered by `options.serverSideRender`, but in a more generic way.

File: lib/middleware.js (hunk 2 above)
Comment (sokra) [reply lib/middleware.js]: `fs` is a in memory filesystem so it doesn't matter

**Comment by evilebottnawi** [comment]
@ferdinando-ferreira 
> sokra [6:10 PM]
> hmm.. not sure if this is needed. It's a middleware, so you could create another middleware to transform the responses. And anyway transformations to assets can be done as webpack plugin...

What do you think about this solution?

**Comment by ferdinando-ferreira** [comment]
> I can propose a smaller change to middleware.js that passes fs along with webpackStats to the next middleware in line 

Done, PR #337. Both PRs solve the same issue, this one in a more generic fashion, the other one contingent on ServerSideRender flag not going away.

**Comment by ferdinando-ferreira** [comment]
Rescinding this PR, #337 is enough as long as ServerSideRequest doesn't get dropped.

**Comment by michael-ciniawsky** [comment]
We will definitely keep it until we found a equally sufficient replacement for SSR workflows. I will review #337 asap, thx for your efforts  on this so far :)

**Pull request closed without merging**